provider/cluster/kube/client_ingress.go

```diff
@@ -0,0 +1,250 @@
+	mtypes "github.com/ovrclk/akash/x/market/types"
+	netv1 "k8s.io/api/networking/v1"
+	kubeErrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/client-go/tools/pager"
+	"strconv"
+	"strings"
+)
+
+func kubeNginxIngressAnnotations(directive cluster.ConnectHostnameToDeploymentDirective) map[string]string {
+	// For kubernetes/ingress-nginx
+	// https://github.com/kubernetes/ingress-nginx
+	const root = "nginx.ingress.kubernetes.io"
+
+	result := map[string]string{
+		fmt.Sprintf("%s/proxy-read-timeout", root): fmt.Sprintf("%dms", directive.ReadTimeout),
+		fmt.Sprintf("%s/proxy-send-timeout", root): fmt.Sprintf("%dms", directive.SendTimeout),
+
+		fmt.Sprintf("%s/proxy-next-upstream-tries", root): strconv.Itoa(int(directive.NextTries)),
+		fmt.Sprintf("%s/proxy-body-size", root):           strconv.Itoa(int(directive.MaxBodySize)),
+	}
+
+	nextTimeoutKey := fmt.Sprintf("%s/proxy-next-upstream-timeout", root)
+	if directive.NextTimeout > 0 {
+		result[nextTimeoutKey] = fmt.Sprintf("%dms", directive.NextTimeout)
+	} else {
+		result[nextTimeoutKey] = "0" // Magic value for disable
+	}
+
+	builder := strings.Builder{}
+
+	for i, v := range directive.NextCases {
+		first := string(v[0])
+		isHTTPCode := strings.ContainsAny(first, "12345")
+
+		if isHTTPCode {
+			builder.WriteString("http_")
+		}
+		builder.WriteString(v)
+
+		if i != len(directive.NextCases)-1 {
+			// The actual separator is the space character for kubernetes/ingress-nginx
+			builder.WriteRune(' ')
+		}
+	}
+
+	result[fmt.Sprintf("%s/next-upstream", root)] = builder.String()
+
+	return result
+}
+
+func (c *client) ConnectHostnameToDeployment(ctx context.Context, directive cluster.ConnectHostnameToDeploymentDirective) error {
+	ingressName := directive.Hostname
+	ns := lidNS(directive.LeaseID)
+	rules := ingressRules(directive.Hostname, directive.ServiceName, directive.ServicePort)
+
+	_, err := c.kc.NetworkingV1().Ingresses(ns).Get(ctx, ingressName, metav1.GetOptions{})
+	metricsutils.IncCounterVecWithLabelValuesFiltered(kubeCallsCounter, "ingresses-get", err, kubeErrors.IsNotFound)
+
+	labels := make(map[string]string)
+	labels[akashManagedLabelName] = "true"
+	appendLeaseLabels(directive.LeaseID, labels)
+
+	obj := &netv1.Ingress{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:        ingressName,
+			Labels:      labels,
+			Annotations: kubeNginxIngressAnnotations(directive),
+		},
+		Spec: netv1.IngressSpec{
+			Rules: rules,
+		},
+	}
+
+	switch {
+	case err == nil:
+		_, err = c.kc.NetworkingV1().Ingresses(ns).Update(ctx, obj, metav1.UpdateOptions{})
+		metricsutils.IncCounterVecWithLabelValues(kubeCallsCounter, "networking-ingresses-update", err)
+	case kubeErrors.IsNotFound(err):
+		_, err = c.kc.NetworkingV1().Ingresses(ns).Create(ctx, obj, metav1.CreateOptions{})
+		metricsutils.IncCounterVecWithLabelValues(kubeCallsCounter, "networking-ingresses-create", err)
+	}
```

Comment:
I know we use this pattern elsewhere, it just struck me that there's a diabolical case:

1. two workers upserting an ingress.
2. both fail on `Update()`, continue on to `Create()`.
3. One `Create()` succeeds, one `Create()` fails*

* Or, possible database conflict/corruption at `3.`?

Reply:
I'll have to look see if two goroutines can be calling this at the same time.

Reply:
It's okay.  We should probably make a generic optimistic locking/retry utility for everywhere that we do this.